### PR TITLE
docs: use moderncmakedomain, Sphinx 4

### PR DESCRIPTION
**Pull request by henryiii**
This enables the use of Sphinx 4. The copy of `docs/cmake.py` from CMake has been replaced with `sphinxcontrib-moderncmakedomain`, which is a port of that file to PyPI, and is much more up-to-date (from CMake 3.19, whereas our copy was much other and didn't support Sphinx 4). Really, it seems like something like that could be maintained as part of scikit-build; curious to know @slurps-mad-rips's opinion on that. We use `sphinxcontrib-moderncmakedomain` in pybind11.

This also pins docutils, as docutils 0.17 is not compatible with sphinx-rtd-theme currently. I'd really like to move away from the rtd theme for this and other reasons.

Requested in #573.

### CODE REVIEW

**Comment by henryiii** [comment]
@jcfr or @thewtex , could you go to ReadTheDocs and activate pull request previews? Admin -> Advanced Settings -> Build pull requests for this project. Then there will be a link to preview the docs in the PR.

**Comment by jcfr** [comment]
>  Admin -> Advanced Settings -> Build pull requests for this project.

All set. To streamline maintenance, could I add you as maintainer on readthedocs ? what is your username ?

**Comment by henryiii** [comment]
I'm `henryiii` almost everywhere (twitter being the only exception I can think of ATM).

**Comment by jcfr** [comment]
Since the push event didn't make it ... I just re-synced the webhook. I am also wondering if the preview would be available only for new PR ?

image

**Comment by henryiii** [comment]
There it does, it's in the list of checks now. It can be a little finicky - resyncing the webhook after enabling it was a good idea.

**Comment by slurps-mad-rips** [comment]
> Really, it seems like something like that could be maintained as part of scikit-build

I'm totally fine with this, and would be willing to transfer ownership/maintainership of the repository and the package on PyPI.

I had originally set it up to help write CMake documentation, but the cmake domain that kitware uses is a bit wonky and expects certain declarations to be in specific locations (at least the last time I tried to use it) and just never ended up using it as a result 😅

**Comment by henryiii** [comment]
Interesting, this had a syntax highlighting problem here: https://scikit-build--574.org.readthedocs.build/en/574/cmake-modules/PythonExtensions.html#example-usage - I noticed the warning when building, but it does look like it's unable to render this CMake snippet, while the current file can.

**Comment by jcfr** [comment]
> noticed the warning when building, but it does look like it's unable to render this CMake snippet

For reference,  here is the warning:

```
/home/jcfr/Projects/scikit-build/skbuild/resources/cmake/FindPythonExtensions.cmake:169: WARNING: Could not lex literal_block as "cmake". Highlighting skipped.
```

**Comment by jcfr** [comment]
> docs: Fix CMake example syntax in FindPythonExtensions documentation

Worth noting that pygment is not aiming to report syntax error and expect the block to be valid.

Quote from https://pygments.org/faq/#can-i-use-pygments-for-programming-language-processing

> Also, error reporting is not the scope of Pygments. It focuses on correctly highlighting syntactically valid documents, not finding and compensating errors.

**Comment by jcfr** [comment]
> I'm totally fine with this, and would be willing to transfer ownership/maintainership of the repository and the package on PyPI

@slurps-mad-rips  Thanks for considering this :pray:  

I am currently discussing with the CMake team and will follow up shortly with an update.

@henryiii  Do  you think  including the update of the `moderncmakedomain` as part of the `cmake-python-distributions` release process something reasonable ? 

In a nutshell, this would consist in copying the `cmake.py` from CMake code base, while would trigger the continuous integration and automatic update to PyPI (based on GitHub Action) that we would be adding to the project.

File: skbuild/resources/cmake/FindPythonExtensions.cmake [hunk 1]
```diff
@@ -201,7 +201,7 @@
 #                            FORWARD_DECL_MODULES_VAR fdecl_module_list)
 #
 #    # module2 -- dynamically linked
-#    include_directories({Boost_INCLUDE_DIRS})
+#    include_directories(${Boost_INCLUDE_DIRS})
```
Comment (henryiii) [inline skbuild/resources/cmake/FindPythonExtensions.cmake:204]: Nice! I was looking for an error, but missed this. :)

**Comment by henryiii** [comment]
> I am currently discussing with the CMake team and will follow up shortly with an update.

That would be great. The "sphinxcontrib-cmakedomain" package seems to be abandoned, so we could ask for that name on PyPI if you wanted something simpler - we could even release to both names for a while. Also, I don't see a "sphinxcontrib-cmake" name, I'm guessing @slurps-mad-rips (and the other package's author) considered that, so maybe there's a reason it's best to avoid just "sphinxcontrib-cmake" here. (?) Anyway, there are existing users, so keeping the same name and just documenting it (or simply having it in the scikit-build org) is fine for discoverability.
 
> @henryiii Do you think including the update of the `moderncmakedomain` as part of the `cmake-python-distributions` release process something reasonable ?
> 
> In a nutshell, this would consist in copying the `cmake.py` from CMake code base, while would trigger the continuous integration and automatic update to PyPI (based on GitHub Action) that we would be adding to the project.

Yes, something like that would be ideal. I'd want it mostly automated, though it could be a separate button for now. Ideally, it would be nice if the CMake releases would trigger the PR creation, which I think from the links you sent earlier is likely possible. Eventually auto-triggering was in the plan when working on the nox tooling.

**Comment by slurps-mad-rips** [comment]
> 
> 
> > I am currently discussing with the CMake team and will follow up shortly with an update.
> 
> That would be great. The "sphinxcontrib-cmakedomain" package seems to be abandoned, so we could ask for that name on PyPI if you wanted something simpler - we could even release to both names for a while. Also, I don't see a "sphinxcontrib-cmake" name, I'm guessing @slurps-mad-rips (and the other package's author) considered that, so maybe there's a reason it's best to avoid just "sphinxcontrib-cmake" here. (?) 

It's considered "well mannered" for sphinx plugins to add `domain` for plugins that add a domain and to not have it present for anything that is not related to a domain. Its not enforced, but I was following de facto best practices that the sphinx community have implicitly agreed upon 🙂

**Comment by henryiii** [comment]
> It's considered "well mannered" for sphinx plugins to add domain

Thats fine, then, not that familiar with sphinx conventions. So it might be worth trying to get the `sphinxcontrib-cmakedomain` name, though, to eventually move it to that? Most maintainers if still around are happy to give up a name if there's a clearly better maintained version of it, and if the maintainer is unreachable, we can request the name from PyPI (it's been dead for 8 years). I can probably set up the setup.py to produce the "other" name on request, so we can dual-release for a while. I do like the term [Modern CMake](https://cliutils.gitlab.io/modern-cmake/), so I don't mind the name moderncmakedomain either :), just thinking `sphinxcontrib-cmakedomain` might be a bit more official and discoverable.

**Comment by slurps-mad-rips** [comment]
They're definitely different in what they do and how they are declared and it could possibly break other folks stuff. That said PyPI didn't have squatting/maintainership policy back when I created the plugin. We'd have to contact the author however and I believe the PyPI folks can help in that regard if necessary, though it can be months before you receive a response of any kind 😅

**Comment by henryiii** [comment]
I don't think anyone is using code from 8 years ago, I'm sure it doesn't support Python 3, etc. And sphinxcontrib-moderncmakedomain has 300x more downloads - probably pretty much only mirrors for the old package. I don't think it should be a blocker, but it might be nice to take that name too and maybe slowly transition if we can. Anyway, let's see what @jcfr and the KitWare team comes back with. :)

**Comment by slurps-mad-rips** [comment]
It appears you would be [right](https://libraries.io/pypi/sphinxcontrib-cmakedomain). It seems there are no *public* dependencies on `cmakedomain` 🙂

**Comment by henryiii** [comment]
I also checked https://pypistats.org/packages/sphinxcontrib-cmakedomain and compared it with https://pypistats.org/packages/sphinxcontrib-moderncmakedomain

**Comment by henryiii** [comment]
And the readme recommends `pip install sphinxcontrib-cmakedomain==0.2`, which would be safe.

**Comment by jcfr** [comment]
:100: 

> Anyway, let's see what @jcfr and the KitWare team comes back with. :)

After sharing our plan with the CMake team, transferring the repo to the `scikit-build` organization is a sensible path forward. 

The authoritative module would still live within the CMake code base and we could update the release process of `cmake-python-distributions`to include the update of the sphinx add-on as a necessary step.

Basically copying the module from the CMake code base, which will trigger CI and automatic update to PyPI associated with `sphinxcontrib-moderncmakedomain` project.

How does that sound ?

**Comment by jcfr** [comment]
> I'd like to get the PyPI ownership swapped out before I transfer the github repository
> as well as who to add as an admin/owner for the pypi package

You could add:
* [henryiii](https://pypi.org/user/henryiii/)
* [jcfr](https://pypi.org/user/jcfr/)
* [kwrobot](https://pypi.org/user/kwrobot/)


> would be easier to then transfer it to scikit-build

As described in the guide [Repository transfers and organizations](https://docs.github.com/en/github/administering-a-repository/managing-repository-settings/transferring-a-repository#repository-transfers-and-organizations)


I just added you to the scikit-build organization, you should now be able to transfer the repository.

image

image

**Comment by slurps-mad-rips** [comment]
> In the process, I also suggest to rename the project `sphinx-moderncmakedomain`

I absolutely missed this, someone else will have to take care of it, sorry. 😔 

The project has been transferred along with an updated `History` section in the README to the scikit-build organization.

PyPI ownership will be added shortly.

**Comment by slurps-mad-rips** [comment]
PyPI invites have been sent out. Once someone has accepted it, I'll remove myself as an owner 🙂

**Pull request merged**